staging/src/k8s.io/code-generator/cmd/validation-gen/lint.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package main
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"k8s.io/gengo/v2/types"
+	"k8s.io/klog/v2"
+)
+
+type linter struct {
+	linted     map[*types.Type]bool
+	rules      []commentRule
+	lintErrors []error
+}
+
+var defaultRules = []commentRule{
+	ruleOptionalAndRequired,
+	ruleRequiredAndDefault,
+}
+
+func (l *linter) AddError(field, msg string) {
+	l.lintErrors = append(l.lintErrors, fmt.Errorf("%s: %s", field, msg))
+}
+
+func newLinter(rules ...commentRule) *linter {
+	if len(rules) == 0 {
+		rules = defaultRules
+	}
+	return &linter{
+		linted:     make(map[*types.Type]bool),
+		rules:      rules,
+		lintErrors: []error{},
+	}
+}
+
+func (l *linter) lintType(t *types.Type) {
+	if _, ok := l.linted[t]; ok {
+		return
+	}
+	l.linted[t] = true
+
+	if t.CommentLines != nil {
+		klog.V(4).Infof("linting type %s", t.Name.String())
+		if err := l.lintComments(t.CommentLines); err != nil {
+			l.AddError("type "+t.Name.String(), err.Error())
+		}
+	}
+	switch t.Kind {
+	case types.Alias:
+		// Recursively lint the underlying type of the alias.
+		l.lintType(t.Underlying)
+	case types.Struct:
+		// Recursively lint each member of the struct.
+		for _, member := range t.Members {
+			klog.V(4).Infof("linting comments for field %s of type %s", member.String(), t.Name.String())
+			if err := l.lintComments(member.CommentLines); err != nil {
+				l.AddError(fmt.Sprintf("field %s.%s", t.Name.String(), member.Name), err.Error())
+			}
+			l.lintType(member.Type)
+		}
+	case types.Slice, types.Array, types.Pointer:
+		// Recursively lint the element type of the slice or array.
+		l.lintType(t.Elem)
+	case types.Map:
+		// Recursively lint the key and element types of the map.
+		l.lintType(t.Key)
+		l.lintType(t.Elem)
+	}
+}
+
+// commentRule is a function that validates a slice of comments.
+type commentRule func(comments []string) error
+
+// lintComments runs all registered rules on a slice of comments.
+func (l *linter) lintComments(comments []string) error {
+	var errs []string
+	for _, rule := range l.rules {
+		if err := rule(comments); err != nil {
+			errs = append(errs, err.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, "; "))
```

Comment:
I think returning a `[]string` for the lint failues, distinct from an error happening, would be clearer and avoid joining and splitting.  Then we can render it like:

```
field Foo.bar.bat: conflicting tags +optional and +required
type Qux: conflicting tags +something and +somethingElse
```

Reply:
Here is assembling the lint error for one type/field. Do you think it still is better to split them though they are for the same type/field?

Reply:
When we output them, we almost certainly want only one per line, right?  If a field has 2 errors (or 3 or 10) we need it to come out in some human-readable form.  We can always join later, but splitting again is ugly.

Reply:
Sounds good. Updated the error output accordingly.